### PR TITLE
lambda functions parameters implemented as Command

### PR DESCRIPTION
**Pull request by pvmm**
This feature adds the power of lambda functions running on cascading events expressed as a simple inline. For instance, this is possible now:

```
CommClient::instance().sendCommand(new CascadingCommand("debug remove_bp bp#3",
        [this] (const QString&) { emit breakpointsChanged(); },    // success
        [this] (const QString& msg) { logError(msg); }             // fail
);
```
and it will run `emit breakpointsChanged();` if `debug remove_bp bp#3` command was successful or `logError(msg);` if it wasn't. Failing lambda function is optional and can be omitted. This feature requires C++11.

### CODE REVIEW

**Comment by m9710797** [comment]
Hi,

In general I'm in favor of this PR. But I have some questions/remarks/suggestions. And most of it is actually critique on the current code rather than on your patch.

* This patch only adds the new functionality but doesn't use it yet. (Great, I like small incremental patches). But can you already outline your plans for this new functionality?
* For example would it be possible to replace (some/most) of the current 'SimpleCommand' uses with 'CascadingCommand'? BTW I'm not a fan of the name 'CascadingCommand', but I also don't have a better suggestion for now (so fine to keep it). Or maybe 'CommandWithCallback'?

The following is critique on the current design, not on this PR. You don't need to address this before I'll accept this PR. But I want to discuss it anyway.
* I don't really like how ownership is currently handled. Every time you see 'delete this' there's something fishy going on.
* Maybe sendCommand should accept a 'unique_ptr<Command>' parameter (instead of 'Command*')?
* Then CommClient becomes responsible for handling lifetime of Command objects, instead of each Command object handling its own lifetime.

(Apart from handling ownership, see above paragraph) Command::cancel() is only really used in ConnectionInfoRequest. I wonder if this cannot be simplified

I see that SimpleCommand is a QObject with a 'replayStatusOk' signal. This signal is only used for 'step_over' and 'step_back'. Would it be possible to replace this signal with the callbacks in the new 'CascadingCommand'? (And then no longer inherit from QObject).

**Comment by pvmm** [comment]
> This patch only adds the new functionality but doesn't use it yet. (Great, I like small incremental patches). But can you already outline your plans for this new functionality?

I am writing a breakpoint viewer, which is basically a table of breakpoints where the user can create, delete and change things about them. Communication with the emulator and between other widgets is basic necessity. So we have to make the communication code simpler and more to the point. This patch is an effort in this direction. The alternative is creating lots of slots and signals, specially if we expect some things to run serially in a code that is by design not serial.

> For example would it be possible to replace (some/most) of the current 'SimpleCommand' uses with 'CascadingCommand'? BTW I'm not a fan of the name 'CascadingCommand', but I also don't have a better suggestion for now (so fine to keep it). Or maybe 'CommandWithCallback'?

Yes, totally doable! How about calling it `CommandObserver`? Let's keep talking. I will answer the remaining comments tonight or the next morning. Probably the latter.

**Comment by m9710797** [comment]
I agree: I like the CascadingCommand alternative better than (many) (small) SimpleCommand subclasses, and certainly better than more slots/signals.

About the naming: what about renaming the current 'Command' class to 'CommandBase' and then use the name 'Command' instead of 'CascadingCommand'?  (Maybe split this over two commits, first the rename, then the addition).

**Comment by pvmm** [comment]
> I don't really like how ownership is currently handled. Every time you see 'delete this' there's something fishy going on.

It seems to be a mixed bag this one. I don't think it is that bad and the language certainly doesn't forbid it in any way.

> Then CommClient becomes responsible for handling lifetime of Command objects, instead of each Command object handling its own lifetime.

I think `shared_ptr` is more appropriate for this, because `shared_ptr` implements reference counting. The creator function has a reference to the Command and the Command itself too. When everyone is done with Command, they just need to call `ptr.reset()`, no explicit de-allocation necessary.

**Comment by pvmm** [comment]
> I see that SimpleCommand is a QObject with a 'replayStatusOk' signal. This signal is only used for 'step_over' and 'step_back'. Would it be possible to replace this signal with the callbacks in the new 'CascadingCommand'? (And then no longer inherit from QObject).

You mean like this?  c886c5f0f1306eaecdaa888a54b7a02c5069bcb3

**Comment by m9710797** [comment]
Thanks. I've merged this PR.
I hope there will be some follow-up PRs that make use of the (new) Command class to cleanup some code.

BTW, you have many PRs open. Thanks a lot for that!
Personally I try to review those that are more technical and/or c++ related.
The PRs that change debugger functionality I leave to Manuel to review.

But because of the amount of PRs it's easily possible that I overlook some. Don't hesitate to send a reminder if you don't get any feedback within a couple of days.

**Comment by m9710797** [comment]
Hi,

I know this PR has been merged. But this still seems like a good thread to continue the discussion.


> > I don't really like how ownership is currently handled. Every time you see 'delete this' there's something fishy going on.
> 
> It seems to be a mixed bag this one. I don't think it is that bad and the language certainly doesn't forbid it in any way.

Yes the language allows it, but that doesn't mean it's a good idea. In my experience there are almost always better alternatives that don't require "delete this".

> > Then CommClient becomes responsible for handling lifetime of Command objects, instead of each Command object handling its own lifetime.
> 
> I think `shared_ptr` is more appropriate for this, because `shared_ptr` implements reference counting. The creator function has a reference to the Command and the Command itself too. When everyone is done with Command, they just need to call `ptr.reset()`, no explicit de-allocation necessary.

I disagree. When unique_ptr is sufficient we should use that instead of shared_ptr.

Compared to shared_ptr, unique_ptr is much more limited, and that's a good thing. When reading the code and you see a unique_ptr, you know that it is the only owner of the object at that point in time. With shared_ptr it might be the only owner, or there might also be others, you don't know. Designs are much easier to understand/reason about when at each point in time there's a clear single owner for each object.

As a bonus (not that important in the context of the debugger), unique_ptr is also easier for the compiler to understand. And thus it can generate simpler/faster code for unique_ptr compared to shared_ptr.

I must admit that I haven't studied the debugger/command code in detail (or at least it's been multiple years). But I think single ownership is sufficient here.

As a more direct reply to your comment:
* Yes, currently the CommandBase keeps itself alive, but a more logical location would be the 'commands' queue in OpenMSXConnection. (So make it: QQueue<std::unique_ptr<CommandBase>> commands).
* Yes there is a creator function for the Command objects. And yes there is the 'OpenMSXConnection::command' queue (currently the Commands themselves). And each have (temporary) ownership of CommandBase objects. But they never have simultaneous ownership. In other words: ownership can be transferred from the creator to the queue. No need for reference counting.

Some examples:
* Sending a command (= creating a command object):
    `comm.sendCommand(std::make_unique<SimpleCommand>("reset")));`
* In the implementation of semdCommand() (=transfer ownership):
 ```
void OpenMSXConnection::sendCommand(std::unique_ptr<CommandBase> command) {
        ....
        commands.enqueue(std::move(command));
```
* No explicit memory management needed anymore, not even reset(). For example
```
     OpenMSXConnection::cancelPending() {
        commands.clear(); // this automatically deletes all pending CommandBase objects
     }
```
In other words: just removing the elements is sufficient (no need for delete, no need for reset()). Similar to how e.g. in a vector<int> just removing the int is sufficient.  (And no need to, behind the scenes, perform reference counting).

--

Anyway, this is a side discussion. It's just an idea. I don't necessarily expect you to work on this. But you're very welcome to do this of course. If you don't pick it up, then maybe I might in the future.

**Comment by pvmm** [comment]
> In other words: just removing the elements is sufficient (no need for delete, no need for reset()). Similar to how e.g. in a vector just removing the int is sufficient. (And no need to, behind the scenes, perform reference counting).

I might be missing something here, but who's to tell the queue to remove the command whose life has ended naturally (not cancelled externally) when only the command itself knows when it should die?

**Comment by m9710797** [comment]
Current situation  (src/OpenMSXConnection.cpp around line 229):
`CommandBase* command = commands.dequeue();`
This removes the command from the queue. Then we either call replyOk() or replyNok(), these instruct the command object to commit suicide.

My proposed situation:
`std::unique_ptr<CommandBase> command = commands.dequeue();`
This removes the command from the queue, and transfers the ownership from the queue to this function.
We then call replyOk() or replyNok(), these do not end the lifetime of the command object.
When the unique_ptr goes out of scope it deletes the object.


Note: my examples were not meant to be complete. E.g. I didn't mention calling the CommandBase::cancel() method.   (But I think we should, orthogonally, restructure the code so that cancel() is no longer needed).

**Comment by m9710797** [comment]
Definitely bump to c++14. Bump immediately to c++17 is also fine.

Feel free to drop my proposal for now.  Or alternatively I can take a look next weekend.
